File: typescript/foxit-pdf-api-mcp-server/src/utils/task-poller.ts
import type { FoxitPDFClient } from "../client";
import type { TaskResponse } from "../types/api";

/**
 * Poll a task until completion or timeout
 */
export async function pollTaskUntilComplete(
  client: FoxitPDFClient,
  taskId: string,
  timeoutMs?: number
): Promise<TaskResponse> {
  const config = client.getConfig();
  const timeout = timeoutMs ?? config.defaultTimeout;
  const pollInterval = config.pollInterval;
  const startTime = Date.now();

  while (true) {
    const taskStatus = await client.getTaskStatus(taskId);

    // Task completed successfully
    if (taskStatus.status === "COMPLETED") {
      return taskStatus;
    }

    // Task failed
    if (taskStatus.status === "FAILED") {
      const error = new Error(
        taskStatus.error?.message ?? "Task failed without error details"
      ) as Error & { code: string; taskId: string; details?: unknown };
      error.code = taskStatus.error?.code ?? "TASK_FAILED";
      error.taskId = taskId;
      error.details = taskStatus.error?.details;
      throw error;
    }

    // Check timeout
    if (Date.now() - startTime > timeout) {
      const error = new Error(
        `Task ${taskId} did not complete within ${timeout}ms`
      ) as Error & { code: string; taskId: string };
      error.code = "TASK_TIMEOUT";
      error.taskId = taskId;
      throw error;
    }

    // Wait before next poll
    await new Promise((resolve) => setTimeout(resolve, pollInterval));
  }
}

/**
 * Execute an operation and wait for completion
 * Returns the final task result
 */
export async function executeAndWait(
  client: FoxitPDFClient,
  operationFn: () => Promise<{ taskId: string }>,
  timeoutMs?: number
): Promise<TaskResponse> {
  const { taskId } = await operationFn();
  return pollTaskUntilComplete(client, taskId, timeoutMs);
}

File: python/foxit-pdf-api-mcp-server/src/foxit_pdf_api_mcp_server/utils/task_poller.py
"""Task polling utilities for async operations."""

import asyncio
from typing import Any, Awaitable, Callable, Mapping, Optional

from ..client.foxit_client import FoxitAPIError, FoxitPDFClient
from ..types.api import TaskResponse


async def poll_task_until_complete(
	client: FoxitPDFClient,
	task_id: str,
	timeout: Optional[int] = None,
) -> TaskResponse:
	"""Poll a task until completion or timeout.

	Args:
		client: Foxit PDF client instance
		task_id: Task ID to poll
		timeout: Timeout in seconds (uses client default if not provided)

	Returns:
		Completed task response

	Raises:
		FoxitAPIError: If task fails or times out
	"""
	timeout_seconds = timeout or client.default_timeout
	poll_interval = client.poll_interval
	start_time = asyncio.get_event_loop().time()

	while True:
		task_status = await client.get_task_status(task_id)

		if task_status["status"] == "COMPLETED":
			return task_status

		if task_status["status"] == "FAILED":
			error_info = task_status.get("error", {})
			error = FoxitAPIError(
				message=error_info.get("message", "Task failed without error details"),
				code=error_info.get("code", "TASK_FAILED"),
				details=error_info.get("details"),
			)
			error.task_id = task_id  # type: ignore[attr-defined]
			raise error

		elapsed = asyncio.get_event_loop().time() - start_time
		if elapsed > timeout_seconds:
			error = FoxitAPIError(
				message=f"Task {task_id} did not complete within {timeout_seconds}s",
				code="TASK_TIMEOUT",
			)
			error.task_id = task_id  # type: ignore[attr-defined]
			raise error

		await asyncio.sleep(poll_interval)


async def execute_and_wait(
	client: FoxitPDFClient,
	operation_fn: Callable[[], Awaitable[Mapping[str, Any]]],
	timeout: Optional[int] = None,
) -> TaskResponse:
	"""Execute an operation and wait for completion."""
	operation_result = await operation_fn()
	task_id = operation_result["taskId"]
	return await poll_task_until_complete(client, task_id, timeout)


__all__ = ["poll_task_until_complete", "execute_and_wait"]
